### ISSUE TITLE
Make PrettyMIDI object serializable efficiently

### ISSUE THREAD

**Issue by bzamecnik** [craffel/pretty-midi#133]
Parsing MIDI is rather slow (eg. `music21` is pretty slow, `pretty-midi` is better, but still not pretty fast) and we might want to perform queries on the parsed data or use that multiple times (eg. as training data for an ML model). Besides trying to optimize the parsing stage another option is to cache the parsed results, eg. by picking or a different form of serialization. In music21 it seems the objects are not serializable at all. In pretty-midi I was able to serialize the `PrettyMIDI` object and load it back, but the problem is that the serialized form several orders bigger than the original MIDI (just prohibitively big - several MB for a few kB of MIDI).

The subject of this issue is to serialize only vital information that can be used to restore the object while keeping any post-processing still faster than parsing the MIDI again.

Originally I though pickle takes derived properties like `get_piano_roll()` abut after a very superficial inspection it seems some internal properties like `__tick_to_time` take much space. I can investigate and measure it in more detail.

The possible solution might be to explicitly provide object for serialization and possibly compress them (eg. dense matrix to sparse) before serialization and decompress after serialization.

The goal is to reduce the pickled size to something comparable to MIDI (or one or two orders bigger) and also to keep the (de)serialization time low.

**Comment by craffel** [comment]
Maybe it would be first to define what you need in terms of speed, memory usage, disk space, etc.  I have done a few projects which involve parsing/analyzing/using ML models on O(100,000) MIDI files with `pretty_midi` on commodity hardware with no issues.

> it seems some internal properties like __tick_to_time take much space.

Yes, this is cached to make things more efficient.  A few megabytes should be no big deal in memory, I think :)

**Comment by cifkao** [comment]
Here is a comparison of creating a `PrettyMIDI` object from MIDI vs. unpickling it. The first file is very small (8 bars of monophony), the second one is a full song.

```pycon
In [1]: %timeit pretty_midi.PrettyMIDI('small.mid')
5.53 ms ± 26.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [2]: %timeit with open('small.pickle', 'rb') as f: pickle.load(f)
227 µs ± 1.26 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [3]: %timeit pretty_midi.PrettyMIDI('smoke.mid')
381 ms ± 17.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [4]: %timeit with open('smoke.pickle', 'rb') as f: pickle.load(f)
23.8 ms ± 265 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Here are the file sizes:
```
 514 small.mid           56K smoke.mid
126K small.pickle       2.1M smoke.pickle
```

**Comment by cifkao** [comment]
@craffel Would it be viable to integrate the [`NoteSequence`](https://github.com/tensorflow/magenta/blob/a33262f5c2fd9f7130de71d8482737fe35fa137b/magenta/protobuf/music.proto#L27) protobuffer from Magenta into `pretty_midi` (either by adding conversion methods or by directly including it as part of the internal representation)? It seems that the proto mimics the design of `pretty_midi`, and the code for converting back and forth [already exists](https://github.com/tensorflow/magenta/blob/a33262f5c2fd9f7130de71d8482737fe35fa137b/magenta/music/midi_io.py#L50). However, it could be inconvenient to use Magenta directly, since it depends on a lot of other packages (e.g. a specific version of TensorFlow).

**Comment by craffel** [comment]
No, I don't think so. `pretty_midi` does not depend or rely on `NoteSequence` in any way; the the dependency graph only points in one direction. If it's hard to use `NoteSequence` because of all of Magenta's dependencies, I'd suggest you advocate for Magenta to factor out `NoteSequence` into a separate library.

**Comment by cifkao** [comment]
[note-seq](https://github.com/magenta/note-seq/) is now a separate library with reduced dependencies, and `NoteSequence` has been [fixed](https://github.com/magenta/note-seq/issues/2) to support efficient pickling!

Now we can do this:
```python
import pretty_midi, note_seq, pickle

pm = pretty_midi.PrettyMIDI('file.mid')

# PrettyMIDI -> NoteSequence -> pickle
ns = note_seq.midi_to_sequence_proto(pm)
with open('file.pickle', 'wb') as f:
    pickle.dump(ns, f)

# pickle -> NoteSequence -> PrettyMIDI
with open('file.pickle', 'rb') as f:
    ns = pickle.load(f)
pm = note_seq.sequence_proto_to_pretty_midi(ns)
```

**Comment by Yikai-Liao** [comment]
Hi @bzamecnik and others in this thread,

This is an interesting discussion, and the points raised about MIDI parsing speed ("not pretty fast") and serialization efficiency being potential bottlenecks, especially for large-scale ML workflows, are very valid.

I'd like to suggest an alternative library, `symusic` ([https://github.com/Yikai-Liao/symusic](https://github.com/Yikai-Liao/symusic)), which I developed specifically with these performance considerations in mind.

**1. Parsing Speed:**
`symusic` leverages a C++ core (based on `minimidi`) and achieves significantly faster parsing speeds – often **200-500x faster** than `mido`-based libraries like `pretty-midi`. This speed can effectively remove the MIDI parsing bottleneck in data preprocessing pipelines for ML.

Here's a benchmark comparison (more details at [symusic-benchmark](https://github.com/Yikai-Liao/symusic-benchmark)):

![Benchmark Chart](https://github.com/user-attachments/assets/5f663e4e-9562-436e-8f97-5b62e96d0314)

In fact, recognizing this performance advantage, the popular MIDI tokenization library `MidiTok` ([https://github.com/Natooz/MidiTok](https://github.com/Natooz/MidiTok)) now uses `symusic` as its backend for reading and writing MIDI files, enabling much faster tokenization workflows for deep learning models.

**2. Serialization Efficiency:**
Furthermore, `symusic` directly addresses the serialization concerns mentioned here regarding large pickled object sizes (often caused by caching internal structures like `__tick_to_time` in `pretty-midi`). `symusic` offers **extremely fast and efficient `pickle` support** out-of-the-box, producing much smaller serialized objects. This significantly speeds up loading preprocessed data, especially when used with multiprocessing, and makes caching a very viable and efficient strategy without excessive disk usage.

**3. Simplified Workflow for ML:**
With its inherent speed and efficient serialization, `symusic` can simplify data pipelines. The need to design or rely on potentially complex or less general intermediate formats (like the discussed `NoteSequence`) primarily for performance reasons might be reduced. You can often work directly with the parsed `symusic` objects throughout your preprocessing and training loops.

It also provides an intuitive note-level API, similar in concept to `pretty-midi`, which might ease the transition.

We recently shared this work and its performance benefits at the **ISMIR 2024 Late-Breaking/Demo session**.

For anyone dealing with performance issues in large-scale MIDI processing or looking for efficient serialization, `symusic` could be a worthwhile alternative.

You can find more details here:
- **GitHub:** [https://github.com/Yikai-Liao/symusic](https://github.com/Yikai-Liao/symusic)
- **Documentation:** [https://yikai-liao.github.io/symusic/](https://yikai-liao.github.io/symusic/)
- **Colab Tutorial:** [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Yikai-Liao/symusic/blob/main/tutorial.ipynb)
- **ISMIR 2024 LBD Abstract:** [https://ismir2024program.ismir.net/lbd_426.html#lbd](https://ismir2024program.ismir.net/lbd_426.html#lbd)

Hope this is helpful!

**Comment by craffel** [comment]
Hi @Yikai-Liao , pretty-midi is just a set of convenience functions built on top of a MIDI parser (currently mido). We changed the MIDI parser in the past from python-midi to mido, we could easily change it again to minimidi if mido is slow; in fact we could even change it to use your symusic library directly (assuming the API you provide covers all of the information pretty_midi needs to build its representation). I'd be happy to discuss such a transition, but I'm not sure it's constructive to post here to suggest users just migrate to your library unless your library covers the full functionality of pretty_midi.
